client/packages/android/app/src/main/java/org/openmsupply/client/ExtendedWebViewClient.java

```diff
@@ -29,35 +31,52 @@ public ExtendedWebViewClient(Bridge bridge) {
     // fails for self signed certificates and plugin definitions etc is not injected
     @Override
     public void onPageStarted(WebView webView, String url, Bitmap favicon) {
+        Logger.debug("onPageStarted" + url   );
         if (url.startsWith("data:text")) return;
 
+        if (this.jsInject == null){
+            this.jsInject = this.generatePluginScript();
+        }
+
         if(this.jsInject != null) {
+            Logger.debug("injecting JS");
             // .post to run on UI thread
             webView.post(() -> webView.evaluateJavascript(this.jsInject, null));
         }
     }
 
+
+    private void sleep(int delay) {
+        try {
+            Thread.sleep(delay);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
     String generatePluginScript() {
         // TODO make sure this is only injected for pages in native bundle
         // There is no way to get the full list of plugins from bridge, use 'debug' and
         // see what plugins to add
-        List<PluginHandle> pluginList = Arrays.asList(
-                bridge.getPlugin("NativeApi"),
-                bridge.getPlugin("Keyboard"),
-                bridge.getPlugin("WebView"),
-                bridge.getPlugin("BarcodeScanner"),
-                bridge.getPlugin("Preferences"),
-                bridge.getPlugin("KeepAwake"),
-                bridge.getPlugin("App"),
-                bridge.getPlugin("Printer")
-        );
+
+        // This function needs to run after plugins are registered, so can't be part of the constructor as order doesn't appear to be consistent.
+        List<String> pluginNames =  Arrays.asList("NativeApi","Keyboard", "WebView","BarcodeScanner","Preferences", "KeepAwake", "App", "Printer");
+        List<PluginHandle> pluginList = new ArrayList<>();
+        for (String pluginName : pluginNames) {
+            PluginHandle plugin = bridge.getPlugin(pluginName);
+            if (plugin == null) {
+                Logger.error("Couldn't find plugin : " + pluginName);
+                continue;
```

Comment:
Not sure about this, Maybe better to keep the old behaviour and crash the app, however would be good if we could somehow present a better error state if this happens again? The `crash` state just shows as loading... Maybe could inject a substitute JS that displays an error?

Reply:
Plugin missing would also cause an error down in the catch statement, was wondering if it's important to know what plugin hasn't been loaded ? 

With above we would get a logic/silent error which are harder to trace right ? (like Android wouldn't know that it's android ? or some other functionality would fail further up the track)

Reply:
I agree a silent error isn't helpful! 
If everything is working right we shouldn't have any plugins missing.
This useful for me to find that sometimes plugins were missing, but not all of them.
I think we should probably either try to display a more useful error like `Plugin XYZ` missing rather than infinite loading?